Update OSS nested tensor docs to focus on NJT

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145402

Updated nested tensor docs to be NJT-centric (instead of NST-centric). They now include:
* High-level description of NST vs. NJT + a recommendation to use NJT
* General NJT construction / usage
* torch.compile() integration w/ dynamic shapes
* Common errors and how to fix them
* Contribution guide
* Data layout / shape information (with diagram)
* Links to more extensive tutorials involving Transformers / SDPA / FlexAttention

cc @cpuhrsch @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ

jbschlosser committed:
Update nested tensor docs to focus on NJT

[ghstack-poisoned]

svekars commented on docs/source/nested.rst:
```diff
@@ -10,213 +10,471 @@ Introduction
+a standard PyTorch tensor interface for applying operations.
+
+A common application of nested tensors is for expressing batches of variable-length sequential data
+present in various domains (e.g. varying sentence lengths, image sizes, audio / video clip lengths).
```
```suggestion
present in various domains (such as varying sentence lengths, image sizes, audio / video clip lengths).
```

soulitzer commented on docs/source/nested.rst:
```diff
@@ -10,213 +10,471 @@ Introduction
+NJT's ``offsets`` data to delineate batch item boundaries within the packed ragged dimension. The
+workaround here is as suggested in the error message.
+
+>>> torch._dynamo.config.capture_scalar_outputs = True
```
Should we also mention `capture_dynamic_output_shape_ops=True` here?

jbschlosser replied:
yeah good point, probably should! added

vadimkantorov commented:
I think it would be valuable to stress that construction of NJT leads to copy of the passed-in tensor list. It's quite obvious given the layout explanations, but still.

Also, a question: does NJT support `values` tensors with some padding to the right? E.g. if shape is set, so that the static-max-sized values/offsets tensors can be used (and the padding on the right be discarded by the kernels)

soulitzer commented on docs/source/nested.rst:
```diff
@@ -10,213 +10,471 @@ Introduction
-+++++++++++++++++++++++++
-
-Even though a NestedTensor does not support ``.size()`` (or ``.shape``), it supports ``.size(i)`` if dimension i is regular.
+In the vein of ``torch.as_tensor``, ``torch.nested.as_nested_tensor`` can be used to preserve autograd
```
This is preexisting, but I'm not sure I want to mention `torch.as_tensor` here. It seems to not actually preserve autograd history in the actual cases you would want to use it, e.g. convert from numpy.

jbschlosser replied:
removed mention of it

soulitzer commented on docs/source/nested.rst:
```diff
@@ -10,213 +10,471 @@ Introduction
+
+NJTs are designed to be used with ``torch.compile()`` for optimal performance. In particular,
+``torch.compile()`` is important to eliminate eager-mode python tensor subclass overhead, which
+affects NJT as it is a python tensor subclass. NJTs work out-of-the-box and
```
I wonder if it is worth mentioning this explicitly, since maybe that is the expectation already when working with compile.

jbschlosser replied:
true, it should be the expectation. I'm making it explicit to be clear about what users should expect from NJT. There are other features within PyTorch that they can't expect to compose with torch.compile without issue, so it's nice to set this one apart via "you should expect a seamless experience with torch.compile"

soulitzer commented on docs/source/nested.rst:
```diff
@@ -10,213 +10,471 @@ Introduction
+Usage with torch.compile
+++++++++++++++++++++++++
+
+NJTs are designed to be used with ``torch.compile()`` for optimal performance. In particular,
```
"NJTs are designed to be used with ``torch.compile()`` for optimal performance."

Could another way of wording this be something like:
- torch.compile can be used to speed up programs using NJTs. This can be especially important for programs involving smaller tensor sizes where the subclass overhead of NJT adds significant part to runtime.

I think it would be less scary for the people who can't use compile, but may otherwise still benefit.

jbschlosser replied:
Part of my reasoning behind making a very bold statement is to shape expectations for performance with NJTs. It's been the case in the past that people use NJTs in eager expecting speedups but fall into the overhead-bound regime and are surprised to find that the result it slower than whatever they were doing before. The idea here was to be very explicit that if you're not using torch.compile, you shouldn't expect NJT to help you.

But I'm also okay with being more nuanced here. The downside is that the dependence on torch.compile is now multi-faceted and it's tricky to make accurate statements about when the user can expect speedups in eager. Consider the padded dense conversion logic utilized behind binary broadcasting and some other ops - in eager, we don't have fusion of the "computation sandwich" and so overhead and memory usage would on par with or worse than manual padded computation in eager.

soulitzer commented on docs/source/nested.rst:
```diff
@@ -10,213 +10,471 @@ Introduction
+>>> output2.shape
+torch.Size([2, j1, 3])
+
+By default, NJTs support
```
Remove "By default"? (not sure we can disable it)

soulitzer reviewed (approved):
Nice work! Added some small comments

jbschlosser commented:
> I think it would be valuable to stress that construction of NJT leads to copy of the passed-in tensor list. It's quite obvious given the layout explanations, but still.

agreed, I think this is important to make clear. I tried to cover this at the end of the Construction section:

```
The above functions all create contiguous NJTs, where a chunk of memory is allocated to store
a packed form of the underlying components (see the Data Layout section below for more
details).
```

but I'll add some text in-line after introducing the first constructor and hopefully this will be clearer

> Also, a question: does NJT support values tensors with some padding to the right? E.g. if shape is set, so that the static-max-sized values/offsets tensors can be used (and the padding on the right be discarded by the kernels) - e.g. combination of static max values storage for the batch, and NJT dynamism inside the values layout

IIUC, yes it does. NJT does this via a combination of `offsets` to specify the beginning of the sequences and `lengths` to indicate the end of sequence data; anything after this is padding. This allows for non-contiguous NJT views of a padded dense tensor.

I think the docs here should eventually cover this case too, but I'm thinking about booting the description of this now until the metadata overhaul has landed. Further, op support for non-contiguous NJTs is lacking comparatively, so I'm hesitant to publicize this until some of those gaps have been closed.

jbschlosser committed:
Update on "Update OSS nested tensor docs to focus on NJT"


Updated nested tensor docs to be NJT-centric (instead of NST-centric). They now include:
* High-level description of NST vs. NJT + a recommendation to use NJT
* General NJT construction / usage
* torch.compile() integration
* Common errors and how to fix them
* Contribution guide
* Data layout / shape information (with diagram)

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
@pytorchbot merge -i

jbschlosser committed:
Update on "Update OSS nested tensor docs to focus on NJT"


Updated nested tensor docs to be NJT-centric (instead of NST-centric). They now include:
* High-level description of NST vs. NJT + a recommendation to use NJT
* General NJT construction / usage
* torch.compile() integration
* Common errors and how to fix them
* Contribution guide
* Data layout / shape information (with diagram)

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "Update OSS nested tensor docs to focus on NJT"


Updated nested tensor docs to be NJT-centric (instead of NST-centric). They now include:
* High-level description of NST vs. NJT + a recommendation to use NJT
* General NJT construction / usage
* torch.compile() integration
* Common errors and how to fix them
* Contribution guide
* Data layout / shape information (with diagram)

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
